nsot/models.py

```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
         }
 
 
+class Circuit(Resource):
+    """Represents two network Interfaces that are connected"""
+
+    # a_endpoint interface
+    a_endpoint = models.OneToOneField(
+        Interface,
+        on_delete=models.CASCADE,
+        db_index=True,
+        null=False,
+        related_name='circuit_a',
+        verbose_name='A-side Interface',
+        help_text='Unique ID of interface at the A-side'
+    )
+
+    # z_endpoint interface
+    z_endpoint = models.OneToOneField(
+        Interface,
+        on_delete=models.CASCADE,
+        db_index=True,
+        null=True,
+        related_name='circuit_z',
+        verbose_name='Z-side Interface',
+        help_text='Unique ID of interface at the Z-side'
+    )
+
+    # We are currently inferring the site_id from the parent (A-side) Interface
+    # in the .save() method
+    site = models.ForeignKey(
+        Site, db_index=True, related_name='circuits',
+        on_delete=models.PROTECT,
+        help_text='Unique ID of the Site this Circuit is under'
+    )
+
+    name = models.CharField(
+        max_length=255, unique=True,
+        help_text='Unique display name of circuit',
+        default='',
+    )
+
+    def __unicode__(self):
+        return u'%s' % self.name
+
+    @property
+    def interfaces(self):
+        intf_list = [self.a_endpoint, self.z_endpoint]
+        intf_list = [i for i in intf_list if i]
+        return intf_list
+
+    @property
+    def addresses(self):
+        interfaces = self.interfaces
+        addresses = [
+            addr
+            for intf in interfaces
+            for addr in intf.addresses.all()
+        ]
+        return addresses
+
+    @property
+    def devices(self):
+        devices = [intf.device for intf in self.interfaces]
+        return devices
+
+    def clean_site(self, value):
+        """Always enforce that site is set."""
+        if value is None:
+            return self.a_endpoint.site_id
+
+        return value
+
+    def clean_a_endpoint(self, value):
```

Comment:
This raises the error, and doesn't allow the circuit to be created. But when you lookup the circuit endpoints it still returns a circuit object. Something is screwy.
```
In [12]: c2 = Circuit.objects.create(a_endpoint=if2, z_endpoint=if1)
...
ValidationError: {u'a_endpoint': u'Interface already used as a z_endpoint'}

In [18]: if1.circuit_z
Out[18]: <Circuit: >

In [19]: if1.circuit_a
Out[19]: <Circuit: foo-bar1:eth0_foo-bar2:eth0>

In [20]: if2.circuit_a
Out[20]: <Circuit: >

In [21]: if2.circuit_z
Out[21]: <Circuit: foo-bar1:eth0_foo-bar2:eth0>

In [36]: c2.z_endpoint.id
Out[36]: 1

In [37]: Circuit.objects.all()
Out[37]: [<Circuit: foo-bar1:eth0_foo-bar2:eth0>]

In [38]: Circuit.objects.get(id=1)
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
<ipython-input-38-063bfb293c29> in <module>()
----> 1 Circuit.objects.get(id=1)

/Users/jathan/sandbox/virtualenvs/nsot/lib/python2.7/site-packages/django/db/models/manager.pyc in manager_method(self, *args, **kwargs)
    125         def create_method(name, method):
    126             def manager_method(self, *args, **kwargs):
--> 127                 return getattr(self.get_queryset(), name)(*args, **kwargs)
    128             manager_method.__name__ = method.__name__
    129             manager_method.__doc__ = method.__doc__

/Users/jathan/sandbox/virtualenvs/nsot/lib/python2.7/site-packages/django/db/models/query.pyc in get(self, *args, **kwargs)
    332             raise self.model.DoesNotExist(
    333                 "%s matching query does not exist." %
--> 334                 self.model._meta.object_name
    335             )
    336         raise self.model.MultipleObjectsReturned(

DoesNotExist: Circuit matching query does not exist.
```